spec/message.abnf

```diff
@@ -0,0 +1,57 @@
+message = [s] *(declaration [s]) body [s]
+
+declaration = let s variable [s] "=" [s] "{" [s] expression [s] "}"
+body = pattern
+     / (selectors 1*([s] variant))
+
+pattern = "{" *(text / placeholder) "}"
+selectors = match 1*([s] selector)
+selector = "{" [s] expression [s] "}"
+variant = when 1*(s key) [s] pattern
+key = nmtoken / literal / "*"
+
+placeholder = "{" [s] (expression / markup / markup-end) [s] "}"
+
+expression = ((literal / variable) [s annotation])
+           / annotation
+annotation = function *(s option)
+option = name [s] "=" [s] (literal / nmtoken / variable)
```

Comment:
What are the expected semantics of an optional value that is a `nmtoken` but not a `name`, e.g. `foo=1`?

Reply:
It parses as a literal, `"1"`. The implementation of `:func` can interpret it as a number if it makes sense to do so for the `foo` option.

Reply:
Hrm, so the value of an option is either a variable or a literal, but the literal can be implicit rather than quoted? Are `{:func foo=|1|}` and `{:func foo=1}` therefore indistinguishable?

Reply:
There's a super-subtle difference between `nmtoken` and `literal` values.

An `nmtoken` value might be validated at parse time and the values that can be present in an `nmtoken` are restricted vs. the values permitted in a literal. The use of numbers is fairly common in existing formatters, Cf. `Intl.NumberFormat` options such as `maximumSignificantDigits`. But other values have limited (and enumerated) values which might be validated at parse time.

A `literal` value probably is a parsing error (invalid argument) when the function wants a number or enumerated value. MF's options are untyped, but the underlying implementation might not be.

There may be a "tripping hazard" here for users who can't see the difference between:

`{:func symbol=US$}` (invalid, as `$` is reserved) and `{:func symbol=|US$|}`

Reply:
> Hrm, so the value of an option is either a variable or a literal, but the literal can be implicit rather than quoted?

Yes.

> Are `{:func foo=|1|}` and `{:func foo=1}` therefore indistinguishable?

As far as I recall we have not had an explicit discussion on this, but my position would be that during formatting the `:func` handler should not be able to distinguish these two from each other.